teps/0123-specifying-on-demand-retry-in-pipelinetask.md

```diff
@@ -0,0 +1,119 @@
+---
+status: proposed 
+title: Specifying on-demand-retry in a PipelineTask
+creation-date: '2022-09-16'
+last-updated: '2022-09-16'
+authors:
+- '@pritidesai'
+---
+
+# TEP-0123: Specifying on-demand-retry in a PipelineTask
+
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases](#use-cases)
+    - [CD Use Case](#cd-use-case)
+    - [CI Use Case](#ci-use-case)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+This TEP proposes a mechanism to allow the users to choose when to retry a failed `pipelineTask` of a pipeline.
+This kind of on demand retry of a `pipelineTask` is executed from scratch with the original `params` and `specifications`.
+The on demand retry of a `pipelineTask` allow the users to continue to retry until the task succeeds. After the offending task
+succeeds, the rest of the `pipeline` continues executing as usual until the completion.
+
+
+## Motivation
+
+The primary motivation of this proposal is to support pipeline authors to overcome any flakes and to allow running
```

Comment:
Why is on demand retry the best way to solve flakes? Why not automated retries? Are there other alternatives?

Reply:
I think one approach (imo) is to create smaller `Pipeline` that are connected through events. That way you can easily re-run a `Pipeline(Run)` (that failed) and it would "continue" the flow

Reply:
yup, creating a smaller pipeline and connect them through events could be an alternative.

But, the disadvantages of this alternative are very painful for the users. I have a user who has deployed one pipeline for 10 different applications. If he has to split up the pipeline for each possible flake, he will end up in a management nightmare of creating additional triggers for each application and managing pieces of pipeline (which might not even make sense if the pieces contains just one single task).

As a user of the pipeline controller, it's reasonable to expect a solution to handle such failures.

Reply:
I agree that smaller pipelines can be an alternative here, but as @pritidesai highlighted, it's not necessarily applicable to all users and all cases. It's a bit like the many small repos vs. large mono-repo divide - we should support users that require large pipelines for their use cases.